java-squid/src/main/java/org/sonar/java/se/ConstraintManager.java

```diff
@@ -0,0 +1,261 @@
+    map.put(variable, result);
+    symbolMap.put(variable.symbol(), result);
+    return result;
+  }
+
+  public SymbolicValue eval(ProgramState programState, Tree syntaxNode) {
+    syntaxNode = skipTrivial(syntaxNode);
+    switch (syntaxNode.kind()) {
+      case NULL_LITERAL: {
+        return SymbolicValue.NULL_LITERAL;
+      }
+      case BOOLEAN_LITERAL: {
+        boolean value = Boolean.parseBoolean(((LiteralTree) syntaxNode).value());
+        if (value) {
+          return SymbolicValue.TRUE_LITERAL;
+        }
+        return SymbolicValue.FALSE_LITERAL;
+      }
+      case VARIABLE: {
+        Symbol symbol = ((VariableTree) syntaxNode).symbol();
+        SymbolicValue result = programState.values.get(symbol);
+        if (result != null) {
+          // symbolic value associated with local variable
+          return result;
+        }
+        break;
+      }
+      case IDENTIFIER: {
+        Symbol symbol = ((IdentifierTree) syntaxNode).symbol();
+        SymbolicValue result = programState.values.get(symbol);
+        if (result != null) {
+          // symbolic value associated with local variable
+          return result;
+        }
+        break;
+      }
+      case LOGICAL_COMPLEMENT: {
+        UnaryExpressionTree unaryExpressionTree = (UnaryExpressionTree) syntaxNode;
+        SymbolicValue val = eval(programState, unaryExpressionTree.expression());
+        if (SymbolicValue.FALSE_LITERAL.equals(val)) {
+          return SymbolicValue.TRUE_LITERAL;
+        } else if (val.equals(SymbolicValue.TRUE_LITERAL)) {
+          return SymbolicValue.FALSE_LITERAL;
+        }
+        // if not tied to a concrete value, create symbolic value with no constraint for now.
+        // TODO : create constraint between expression and created symbolic value
+      }
+    }
+    return createSymbolicValue(syntaxNode);
+  }
+
+  /**
+   * Remove parenthesis and type cast.
+   */
+  private static Tree skipTrivial(Tree expression) {
+    do {
+      switch (expression.kind()) {
+        case PARENTHESIZED_EXPRESSION:
+          expression = ((ParenthesizedTree) expression).expression();
+          break;
+        case TYPE_CAST:
+          expression = ((TypeCastTree) expression).expression();
+          break;
+        default:
+          return expression;
+      }
+    } while (true);
+
+  }
+
+  public boolean isNull(ProgramState ps, SymbolicValue val) {
+    return NullConstraint.NULL.equals(ps.constraints.get(val));
+  }
+
+  public boolean isConstrained(ProgramState ps, SymbolicValue val) {
+    return ps.constraints.containsKey(val);
+  }
+
+  public Pair<ProgramState, ProgramState> assumeDual(ProgramState programState, Tree condition) {
+    // FIXME condition value should be evaluated to determine if it is worth exploring this branch. This should probably be done in a
+    // dedicated checker.
+    condition = skipTrivial(condition);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Introduce a new variable instead of reusing the parameter "condition". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1226)